runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapperTest.java

```diff
@@ -626,13 +650,29 @@ public void testSourceWithNoReaderDoesNotShutdown() throws Exception {
 
       try {
         thread.start();
-        List<UnboundedSource.UnboundedReader<KV<Integer, Integer>>> localReaders =
-            sourceWrapper.getLocalReaders();
-        while (localReaders != null && !localReaders.isEmpty()) {
-          Thread.sleep(200);
-          // should stay alive
-          assertThat(thread.isAlive(), is(true));
+        // Wait to see if the wrapper shuts down immediately in case it doesn't have readers
+        if (!shouldHaveReaders) {
+          // The expected state is for finalizeSource to sleep instead of exiting
+          while (true) {
+            StackTraceElement[] callStack = thread.getStackTrace();
+            if (callStack.length >= 2
+                && "sleep".equals(callStack[0].getMethodName())
+                && "finalizeSource".equals(callStack[1].getMethodName())) {
+              break;
+            }
+            Thread.sleep(10);
+          }
+        }
+        // Source should still be running even if there are no readers
+        assertThat(sourceWrapper.isRunning(), is(true));
+        synchronized (checkpointLock) {
+          // Trigger emission of the watermark by updating processing time.
+          // The actual processing time value does not matter.
+          sourceWrapper.onProcessingTime(42);
         }
+        // Source should still be running even when watermark is at max
+        assertThat(sourceWrapper.isRunning(), is(true));
+        assertThat(thread.isAlive(), is(true));
         sourceWrapper.cancel();
       } finally {
         thread.interrupt();
```

Comment:
@mxm wouldn't thread.join wait forever and the test timeout when an exception occurs? Maybe we should change that so that exceptions are not swallowed/hidden? I did not see any exception in the log for the array length..

Reply:
Did that also! Now an exception thrown from the test thread will show up in the log/test report.